### PR TITLE
scaffolder: Add support for granting access on create

### LINKED ISSUE #2311: Scaffolder: Automatically add access to scaffolded repos
## Feature Suggestion

Currently scaffolder doesn't explicitly add access to anyone when it creates a repository.
Only org owners / the owner of the PAT token (in the case of GitHub) will have access to it.

This doesn't scale very well.

Bring your own PAT isn't a good solution either as in many cases users won't have access to create repositories directly.

From our perspective it would be ideal to be able to specify a team that should get access to the repo.

Likely it's very organisation specific so it would be good if there's a way to provide your own specific implementation.

At the very minimum to make this work I think if the user who created the repo was given access then they could setup the permissions from there (but not ideal).

## Context

Someone who has minimal access is able to create a repo for their team to work on and will have the appropriate access

### PR DESCRIPTION
**Pull request by timja**
## Hey, I just made a Pull Request!

Fixes https://github.com/spotify/backstage/issues/2311

image

I think scaffolder UI really needs to be dependent on the SCM where the repo is stored.

Many things can be added later:

Pulling orgs from what the token / (GHApp in the future)
Adding teams to an autocomplete
Multiple teams access? (debatable as once one person / team has access they can just add more)
Permission level - currently hardcoded to admin, needs feedback from others if more is needed here

This PR is enough for us to be able to use scaffolder in production. So would be great to get it in.

#### :heavy_check_mark: Checklist

- [x] All tests are passing `yarn test`
- [x] Screenshots attached (for UI changes)
- [x] Relevant documentation updated
- [x] Prettier run on changed files
- [x] Tests added for new functionality
- [x] Regression tests added for bug fixes

### CODE REVIEW

File: plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx
Comment (timja) [inline plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx:57]: wondering if this should be optional as other providers might not need this?

**Comment by timja** [comment]
@benjdlambert would you be able to take a look at this one too please?

**Comment by benjdlambert** [comment]
@timja yep will do, the whole core team is a little busy today, but will see if I can get to it!

File: plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx
Comment (benjdlambert) [reply plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx:57]: yep, I think leave this as optional for now/

**Comment by benjdlambert** [comment]
Hey @timja this looks good up front, want to have chat with @spotify/backstage-core tomorrow morning about if we can use some of the groups in the software catalog to reference the github and other potential groups to create canonical references to groups and see what the status of that work is. Will get back to you as it might be an easier way to do this rather than free text.

**Comment by timja** [comment]
> Hey @timja this looks good up front, want to have chat with @spotify/backstage-core tomorrow morning about if we can use some of the groups in the software catalog to reference the github and other potential groups to create canonical references to groups and see what the status of that work is. Will get back to you as it might be an easier way to do this rather than free text.

Sure this was just an effort to unblock now in a small way.
Fully expecting it to evolve to be a lot better.

Right now it's not really usable without some way of adding access

File: plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts:58]: Would like some documentation about this functionality

File: plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts:60]: What happens if this repo is not in an organisation?

File: plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts:58]: Only do this if the access field is provided

File: plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx
Comment (benjdlambert) [inline plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx:73]: maybe default to `owner`

**Review by benjdlambert — CHANGES_REQUESTED** [review]
OK, I've got a few questions here and some recommended changes.

I will chat to the team tomorrow, but I'm not sure that the required stuff is ready to go our side yet for the groups part. So if we haven't we still have the option to merge this.

File: plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts
Comment (timja) [reply plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts:60]: Only teams come through this flow, org/team.

I can add this check if you think it's worth it?

`user.data.type === 'Organization'`

but it will only happen by accident and don't think we have the validation implemented yet above json-schema so can't really feed back to the client right now.

File: plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts
Comment (timja) [reply plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts]: what / where?

It's documented in the description on the scaffolder page.

Does it need more than the help text?

File: plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts
Comment (benjdlambert) [reply plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts:60]: Hmmm, 
Sorry I thought that the difference in this ternary might make this operation fail no?
https://github.com/spotify/backstage/blob/7806decbc4e19f4338600695a177efdea24f1e87/plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts#L51-L58

File: plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts
Comment (timja) [reply plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts]: https://github.com/spotify/backstage/pull/2442/commits/0c7acbe3cbc7f700fede8bad9ac057c08db2a052

File: plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts
Comment (timja) [reply plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts:60]: Individual users can't have teams, so the access.match should only be validating against an individual user.

i.e. this case shouldn't happen unless a user is trying it to see what happens, or accidentally is trying to create it in their account.

Comment (benjdlambert) [reply plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts:60]: ah ok, now I understand. Sweet.

File: plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts:67]: Condense this into an `else if`

File: plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx
Comment (benjdlambert) [inline plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx:70]:
```suggestion
    accessList: {
```

File: plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx
Comment (benjdlambert) [inline plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx:73]:
```suggestion
      description: 'Who should have access, in org/team or user format. To add multiple, use (,) to seperate',
```

File: plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts:47]: This will change to `accessList`

**Review by benjdlambert — CHANGES_REQUESTED** [review]
Last ones I think

File: plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx
Comment (timja) [reply plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx:73]: so you want support for multiple added here? I wasn't sure if required in the beginning.

I'm not really sure the value of adding multiple teams with admin permission.

Normally we would have an admins group and a group with write access, I didn't build this here as adding permission to the UI would make it more complicated in the initial PR

Ideally it would be something like this built but I didn't want to do it in this pull request:

image

Comment (benjdlambert) [reply plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx:73]: Shit sorry, ignore me.

I've misread `const [, team] = access.split('/')` to think we were splitting on commas. 😐

It's been a long day. This looks good, Will merge.

Your recommendation for multiple is good and I think we will make something like this. But will chat tomorrow with the team for next steps.

**Review by benjdlambert — APPROVED** [review]

**Comment by timja** [comment]
@Rugvip did you want to take a look?

/ merge? 😄

File: plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts
Comment (Rugvip) [inline plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts:83]: Would ``access?.startsWith(`${owner}/`)`` work here? or `.includes()`

**Review by Rugvip — APPROVED** [review]
Yep, good stuff! One small optional extra :grin:

File: plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts
Comment (timja) [reply plugins/scaffolder-backend/src/scaffolder/stages/publish/github.ts]: Done, raised https://github.com/spotify/backstage/pull/2492 as when I went to retest this master wasn't working.

**Pull request merged**
